LC

anthezium opened the pull request:
This branch has three sets of complementary changes:
- A new pass LambdaCaseToLC to produce output that lcc can consume, a corresponding --Sc flag for alb, and a script to glue alb to lcc (hbc.sh) that assumes lcc lives at ../mil/.
- Updates to the build infrastructure to enable building with stack, and other improvements (including a clean target).
- A fib example in examples/ that builds with lcc and its dependencies, copied from tests.

make test currently fails, I don't know whether or not this is a new issue.

anthezium committed:
typo

anthezium committed:
albc has some extra dependencies that weren't
recorded in the cabal file.  clearly we don't use cabal
to build the executables, but adding the dependencies in
a section for albc seems like the best way to document
what actually depends on these packages.

anthezium committed:
stack and hpc generate a bunch of dotfiles.

anthezium committed:
Builds with stack lts-6.6, falls back to standard build
unless you make stack first.  Make nostack turns stack build
off.

anthezium committed:
Habit example fib.hb for testing.

anthezium committed:
Glue alb to lcc

anthezium committed:
Hack away things like (Bit (Nat 32))

anthezium committed:
Add --Sc flag to output LC code suitable for compiling with lcc.
Uses new lambdaCaseToLC to pass.

anthezium committed:
lcc needs operator names to look like symbol$ident

anthezium committed:
Output appears to work with mil (lcc) commit
491041db254a5b81656e002fae8140b71212efa9

Removed e.g. True to TRUE mapping on type/constructor/tylabel names.

Cleaned up cruft.

anthezium committed:
expand all target to hit all executables

add clean target to remove all built stuff and cabal/stack caches

anthezium committed:
retab

anthezium committed:
Merge branch 'master' of github.com:habit-lang/alb into lc

It appears that we need indentation >= 0.2.1

anthezium commented:
Although this patch generally maintains the behavior of other passes, the change in Fidget.RenameTypes will affect the output of the ToThunkified (and therefore also ToFidgetted) command line switches, by no longer occasionally replacing types like (Bit 32) by (Bit (Nat 32)).  It may be worth taking a look at the changes to src/Fidget/RenameTypes.hs and thinking about whether they may have unintended consequences.

jgbm commented:
Looks good; two comments.

While I'm not sure there's any use of the Fidget pipeline at the moment, it seems a pity to break it unnecessarily.  I wonder if it'd be possible to separate a generic RenameTypes pass (i.e., the part that doesn't replace 32 by (Nat 32)) and a Fidget-specific MungeNats pass or the like, and then only invoke the first pass when doing the LC translation.

Also, a silly point.  I had forgotten that I'd left the albc source in the tree; my goal had been to move that kind of thing into the alb driver.  In particular, that avoids having a profusion of .whatsit files.  So I wonder if you couldn't follow a similar pattern to the 'CompCert' pass to invoke the MIL back-end directly from alb instead of via scripts.

anthezium committed:
Invoke lcc from alb using --lcc option.
By default, root for lcc is dir_containing_alb/../mil,
can change it with the --lcc-root option,
can add other command line options with --lcc-other,
can just print lcc command without invoking it with --fake-lcc.

Remove hbc.sh, no longer needed.

anthezium committed:
Copy src/Fidget/RenameTypes.hs to src/LC/RenameTypes.hs, preserving
change to nix (Bit (Nat 32)) in the latter and reverting it in the
former.

The LambdaCase compilation pipeline now produces the code it originally
did, while the LC pipeline produces types without the Nat constructor.

anthezium commented:
For now there's some code duplication, but the Fidget passes generate the code they used to, and the LC passes generate code without e.g. Bit (Nat 32).  I'll refactor later on.

Removed hbc.sh, added an --lcc flag and friends to invoke lcc from alb, just like the CompCert pass.
